metagraph_numba/compiler.py

```diff
@@ -1 +1,320 @@
+        counter = prefix + "_counter"
+        value = getattr(self, counter)
+        setattr(self, counter, value + 1)
+        return f"{prefix}{value}"
+
+    def register_var(self, key):
+        """Register an external values associated with `key`.
+        
+        Returns uniquely generated symbol name.
+        """
+        var_sym = self.next_var()
+        self.var_sym_to_key[var_sym] = key
+        self.var_key_to_sym[key] = var_sym
+        return var_sym
+
+    def register_const(self, value):
+        """Register a Python object as a compile time constant.
+
+        Returns uniquely generated symbol name.
+        """
+        const_sym = self.next_const()
+        self.const_sym_to_value[const_sym] = value
+        return const_sym
+
+    def find_symbol(self, arg):
+        """Find the symbol associated with a given function argument.
+
+        If the argument is a known key for either an external value
+        or a previously registered function call, the appropriate symbol
+        (either `varN` or `retN`) for the value will be returned.
+
+        If this argument is not a known key, return None
+        """
+        if ishashable(arg):
+            # is this a Dask key that maps to an input variable?
+            sym = self.var_key_to_sym.get(arg, None)
+            if sym is not None:
+                return sym
+
+            # is this a Dask key that maps to a function?
+            func_sym = self.func_key_to_sym.get(arg, None)
+            if func_sym is not None:
+                # the symbol is the return value from calling the function
+                return self.func_sym_to_ret_sym[func_sym]
+
+        # return None means no symbol found
+
+    def register_func(self, key, func, args):
+        """Register a function call with return value associated with key and
+        the given argument list.
+
+        The function symbol and return value symbol will be generated and
+        returned. Arguments will be scanned and logged as `var` or `ret`
+        symbols (if they are known Dask keys), or the argument value will be
+        stored and a newly generated const symbol recorded in the argument
+        list.
+
+        Note that function calls must be registered in topologically sorted
+        dependency order, or the key discovery will not work.
+        """
+        func_sym = self.next_func()
+        ret_sym = self.next_ret()
+        self.func_sym_to_key[func_sym] = key
+        self.func_key_to_sym[key] = func_sym
+        self.func_sym_to_func[func_sym] = func
+        self.func_sym_to_ret_sym[func_sym] = ret_sym
+
+        arg_sym_list = []
+        for arg in args:
+            arg_sym = self.find_symbol(arg)
+            if arg_sym is None:
+                arg_sym = self.register_const(arg)
+            arg_sym_list.append(arg_sym)
+
+        self.func_sym_to_args_sym[func_sym] = tuple(arg_sym_list)
+
+        return func_sym, ret_sym
+
+
+def construct_call_wrapper_text(
```

Comment:
Generating the code via string manipulation seems like a good solution for now, but if the code gets any more complicated, perhaps we should consider using Python's AST library instead. 

https://greentreesnakes.readthedocs.io/en/latest/tofrom.html

Reply:
Good point.  Code generation via AST is an interesting idea.  We don't have a lot of experience with AST manipulation from Numba because of historical reasons, but would be a good thing to learn about at some point.